Documentation: Add graceful restart section in BGP documentation (#26354)
cilium-team/cilium: Documentation/network/bgp-control-plane.rst

   Change of an existing neighbor configuration can cause reset of the existing BGP
   peering connection, which results in route flaps and transient packet loss while
   the session reestablishes and peers exchange their routes. To prevent packet loss,

>    it is recommended to configure BGP graceful restart.

Author: @YutaroHayakawa @rastislavs Is this recommendation okay to give in the document?

Reviewer: yeah, makes sense to me :)

Graceful Restart
''''''''''''''''
The Cilium BGP control plane can be configured to act as a graceful restart 
``Restarting Speaker``. By enabling graceful restart, BGP session will restart and

> graceful restart capability will be advertised in the BGP OPEN message.

Reviewer 2: ```suggestion
``Restarting Speaker``. When you enable graceful restart, the BGP session restarts and
the "graceful restart" capability will be advertised in the BGP OPEN message.
```

> In the event of a cilium agent restart, the peering BGP router will not withdraw 
> routes received from the Cilium BGP control plane immediately. As the datapath
> continues to forward traffic during agent restart, there will be no traffic
> disruption. 

Reviewer 2: ```suggestion
In the event of a Cilium Agent restart, the peering BGP router does not withdraw 
routes received from the Cilium BGP control plane immediately. The datapath
continues to forward traffic during Agent restart, so there is no traffic
disruption.
```

> The graceful restart capability is configured on per neighbor basis like this:

Reviewer 2: ```suggestion
Configure graceful restart on a per-neighbor basis, as follows:
```

.. code-block:: yaml

   apiVersion: "cilium.io/v2alpha1"
   kind: CiliumBGPPeeringPolicy
   #[...]
   virtualRouters: # []CiliumBGPVirtualRouter
    - localASN: 64512
      # [...]
      neighbors: # []CiliumBGPNeighbor
       - peerAddress: 'fc00:f853:ccd:e793::50/128'
         # [...]
         gracefulRestart:
           enabled: true
           restartTimeSeconds: 120

> .. note::
>    When enabled, graceful restart capability is advertised for IPv4 and IPv6 address families.

Reviewer 2: ```suggestion
.. note::

   When enabled, graceful restart capability is advertised for IPv4 and IPv6 address families.
```

> Users also have an option to set ``RestartTime`` parameter. ``RestartTime`` is the time
> advertised to the peer within which Cilium BGP control plane is expected to re-establish
> the BGP session after a restart. On expiration of ``RestartTime``, peer will remove 
> previously advertised routes by Cilium BGP control plane.

Reviewer 2: ```suggestion
Optionally, you can use the ``RestartTime`` parameter. ``RestartTime`` is the time
advertised to the peer within which Cilium BGP control plane is expected to re-establish
the BGP session after a restart. On expiration of ``RestartTime``, the peer removes 
the routes previously advertised by the Cilium BGP control plane.
```

> When the Cilium agent restarts, TCP FIN will be generated as BGP TCP socket is closed. Peer
> on receiving TCP FIN will change BGP state to ``Idle`` and ``RestartTime`` timer will start.

Reviewer 2: ```suggestion
When the Cilium Agent restarts, it closes the BGP TCP socket, causing the emission of a TCP FIN packet. On receiving this TCP FIN, the peer changes its BGP state to ``Idle`` and starts its ``RestartTime`` timer.
```

The Cilium agent boot up time varies depending on the deployment. Users are advised to 

> configure ``RestartTime`` greater than time taken by the Cilium agent to boot up. 

Reviewer 2: ```suggestion
The Cilium agent boot up time varies depending on the deployment. If using ``RestartTime``, you should set it to a duration greater than the time taken by the Cilium Agent to boot up.
```
